infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/TopologyDescriptor.java

```diff
@@ -48,81 +70,80 @@ public class TopologyDescriptor {
-                sourceClusters.put(cluster.getClusterId(), cluster);
-            } else if (clusterConfig.getRole() == ClusterRole.SINK) {
-                sinkClusters.put(cluster.getClusterId(), cluster);
-            } else {
-                invalidClusters.put(cluster.getClusterId(), cluster);
-            }
-        }
-
-        setLocalDescriptor(localNodeId);
-    }
 
     /**
      * Constructor
      *
-     * @param topologyConfigId topology configuration identifier (epoch)
-     * @param sourceClusters source cluster's
-     * @param sinkClusters sink cluster's
+     * @param topologyConfigId  topology ID
+     * @param localNodeId      the identifier of this node
+     * @param remoteSinkToReplicationModel  remote Sink to local cluster and corresponding replication models
+     * @param remoteSourceToReplicationModel  remote Source to local cluster and corresponding replication models
+     * @param remoteClusterEndpoints remote clusters to which the local cluster will initiate connection
+     * @param otherClusters      remote clusters which are neither Source nor Sink to the local cluster
      */
-    public TopologyDescriptor(long topologyConfigId, @NonNull List<ClusterDescriptor> sourceClusters,
-                              @NonNull List<ClusterDescriptor> sinkClusters, String localNodeId) {
+    public TopologyDescriptor(long topologyConfigId, String localNodeId,
+                              Map<ClusterDescriptor, Set<LogReplication.ReplicationModel>> remoteSinkToReplicationModel,
+                              Map<ClusterDescriptor, Set<LogReplication.ReplicationModel>> remoteSourceToReplicationModel,
+                              Set<ClusterDescriptor> remoteClusterEndpoints, Set<ClusterDescriptor> otherClusters) {
+
         this.topologyConfigId = topologyConfigId;
-        this.sourceClusters = new HashMap<>();
-        this.sinkClusters = new HashMap<>();
-        this.invalidClusters = new HashMap<>();
+        this.remoteSinkClusterToReplicationModels = remoteSinkToReplicationModel;
+        this.remoteSourceClusterToReplicationModels = remoteSourceToReplicationModel;
+        remoteClusterEndpoints.forEach(cluster -> this.remoteClusterEndpoints.put(cluster.getClusterId(), cluster));
+
+        this.remoteSinkClusters = new HashMap<>();
+        remoteSinkToReplicationModel.keySet().stream().forEach(cluster -> {
+            remoteSinkClusters.put(cluster.getClusterId(), cluster);
+            allClustersInTopology.put(cluster.getClusterId(), cluster);
+        });
+
+
+        this.remoteSourceClusters = new HashMap<>();
+        remoteSourceToReplicationModel.keySet().stream().forEach(cluster -> {
+            remoteSourceClusters.put(cluster.getClusterId(), cluster);
+            allClustersInTopology.put(cluster.getClusterId(), cluster);
+        });
+
+        otherClusters.stream().forEach(cluster -> {
+            allClustersInTopology.put(cluster.getClusterId(), cluster);
```

Comment:
Any specific reason for doing `forEach` on a stream()? I noticed we are using mixed patterns in this file, but the traditional `forEach` on Iterable should be enough?